synthesizer/src/vm/mod.rs

```diff
@@ -109,81 +112,71 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
-                }
-            }
-
-            // Once all the imports have been included, add the parent deployment.
-            deployments.push((*program_id, deployment));
-
-            Ok(deployments)
-        }
-
         // Retrieve the transaction store.
         let transaction_store = store.transaction_store();
+        // Retrieve the block store.
+        let block_store = store.block_store();
         // Retrieve the list of deployment transaction IDs.
         let deployment_ids = transaction_store.deployment_transaction_ids().collect::<Vec<_>>();
-        // Load the deployments from the store.
+        // Initialize storage for the deployments and block heights.
+        let mut deployments = HashMap::with_capacity(deployment_ids.len());
+        let mut heights = Vec::with_capacity(deployment_ids.len());
+        // Load the deployments and their block heights from the store.
         for (i, chunk) in deployment_ids.chunks(256).enumerate() {
             debug!(
                 "Loading deployments {}-{} (of {})...",
                 i * 256,
                 ((i + 1) * 256).min(deployment_ids.len()),
                 deployment_ids.len()
             );
-            let deployments = cfg_iter!(chunk)
-                .map(|transaction_id| {
-                    // Load the deployment and its imports.
-                    load_deployment_and_imports(&process, transaction_store, **transaction_id)
-                })
-                .collect::<Result<Vec<_>>>()?;
-
-            for (program_id, deployment) in deployments.iter().flatten() {
-                // Load the deployment if it does not exist in the process yet.
-                if !process.contains_program(program_id) {
-                    process.load_deployment(deployment)?;
+            // Load the deployments and their block heights.
+            let chunk_iter = cfg_iter!(chunk).map(|transaction_id| {
+                // Retrieve the deployment from the transaction ID.
+                let deployment = match transaction_store.get_deployment(transaction_id)? {
+                    Some(deployment) => deployment,
+                    None => bail!("Deployment transaction '{transaction_id}' is not found in storage."),
+                };
+                // Retrieve the height.
+                let height = match block_store
+                    .find_block_hash(transaction_id)?
+                    .map(|hash| block_store.get_block_height(&hash))
+                {
+                    Some(Ok(Some(height))) => height,
+                    _ => {
+                        bail!("Block height for deployment transaction '{transaction_id}' is not found in storage.")
+                    }
+                };
+                // Return the deployment and height.
+                let transaction_id = transaction_id.clone();
+                let local_map = HashMap::from([(*transaction_id, deployment)]);
+                let local_vec = vec![(*transaction_id, height)];
```

Comment:
why the extra allocations for each deployment and height? can't we just use `push` and `insert` for individual items respectively in the `reduce` block?

Reply:
Would I need to share the underlying `HashMap` and `Vec` need to be shared across threads? 
I was originally trying to avoid lock contention.

I did consider a solution where I collect the deployments and heights and then add it into the `HashMap` and `Vec` serially, but it turned out to slightly slower on [this](https://github.com/ProvableHQ/aleo-program-regressions/tree/master/regressions/profile_process_load_from_store)

Reply:
no, you should be able to return individual objects from this iterator; you would only need to change the later calls to `extend` to `push` and `insert`